src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java

```diff
@@ -29,45 +33,135 @@
+			assertFalse("rendezvous connection should be closed", rendezvousConnection.isOpen());
+			return clientConnection;
+		}).thenCompose(clientConnection -> {
+			return CompletableFutureUtil.delayAsync(Duration.ofMillis(waitMS), EXECUTOR).whenComplete(($void, ex) -> {
+				assertFalse("client connection should be closed implicitly when rendezvous connection closes", clientConnection.isOpen());
+			});
+		});
+		
+		CompletableFuture.allOf(clientClosing, rendezvousClosing).whenComplete(($void, ex) -> {
+			if (ex != null) {
+				fail(ex.getMessage());
+			}	
+			listener.close();
+		}).join();
 	}
 	
 	@Test
-	public void acceptWebSocketConnectionTest() {
-		CompletableFuture<Boolean> checkSocketConnectionTask = new CompletableFuture<Boolean>();
-		CompletableFuture<HybridConnectionChannel> conn = listener.acceptConnectionAsync();
-		CompletableFuture<HybridConnectionChannel> clientConnectionTask = client.createConnectionAsync();
-		conn.thenAccept((connection) -> {
-			checkSocketConnectionTask.complete(true);
-			clientConnectionTask.thenAccept(clientConnection -> clientConnection.closeAsync());
+	public void acceptHandlerTest() throws InterruptedException, ExecutionException, TimeoutException {
+		HybridConnectionListener listener = new HybridConnectionListener(CONNECTION_URI, tokenProvider);
+		AtomicInteger handlerExecuted = new AtomicInteger(0);
+		listener.openAsync().join();
+		
+		// Handler accepts the connection
+		listener.setAcceptHandler(context -> {
+			handlerExecuted.incrementAndGet();
+			return true;
 		});
-		assertTrue("Listener failed to accept connections from sender in webSocket mode.", checkSocketConnectionTask.join());
+		
+		client.createConnectionAsync().whenComplete((clientConnection, ex) -> {
+			assertTrue("clientConnection should have been accepted", clientConnection != null && clientConnection.isOpen());
+		});
+		listener.acceptConnectionAsync().whenComplete((rendezvousConnection, ex) -> {
+			assertTrue("rendezvousConnection should have been accepted", rendezvousConnection != null && rendezvousConnection.isOpen());
+			rendezvousConnection.closeAsync();
+		}).join();
+		
+		// Handler rejects the connection
+		AtomicReference<RelayedHttpListenerContext> httpContext = new AtomicReference<RelayedHttpListenerContext>();
+		listener.setAcceptHandler(context -> {
+			httpContext.set(context);
+			handlerExecuted.incrementAndGet();
+			return false;
+		});
+		CompletableFuture<HybridConnectionChannel> acceptConnectionTask = listener.acceptConnectionAsync();
+		
+		client.createConnectionAsync().handle((clientConnection, ex) -> {
+			assertNull("clientConnection should not have been accepted", clientConnection);
+			assertNotNull("Rejection should have thrown.", ex);
+
+			RelayedHttpListenerResponse response = httpContext.get().getResponse();
+			assertEquals("The rejection response code is incorrect", HttpStatus.BAD_REQUEST_400, response.getStatusCode());
```

Comment:
>HttpStatus.BAD_REQUEST_400, response.getStatusCode() [](start = 60, length = 52)

Is there any way to verify the error details that went over the wire instead of sneaking the internal object from the Listener to the Sender and checking its values?

Reply:
don't think so, the websocket mode doesn't deal with the http upgrade response.

---
In reply to: [261786150](https://github.com/Azure/azure-relay-java/pull/28#discussion_r261786150) [](ancestors = 261786150)

Reply:
The current approach bothers me. This code is checking default assigned values on an object which I'm not sure it's really allowed to be holding on to at this point.

---
In reply to: [261800909](https://github.com/Azure/azure-relay-java/pull/28#discussion_r261800909) [](ancestors = 261800909,261786150)

Reply:
Iteration 3 is wonderful!

---
In reply to: [262212380](https://github.com/Azure/azure-relay-java/pull/28#discussion_r262212380) [](ancestors = 262212380,261800909,261786150)